### DISCUSSION TITLE
I2C sensors

### DISCUSSION THREAD

**Discussion by nagyrobi** [HASwitchPlate/openHASP#60] [Ideas]
Would it be possible to add software support for example SHTC3 (high precision temperature and humidity sensor) or MAX44009 (ambient light sensor) both on I2C bus. All it needs to do is publish to an MQTT topic the value of the sensors at regular time interval, say, at every 30 seconds.

I'm specifically looking at Lanbon L8. It has an analog temp sensor by factory, but that kind is quite unreliable and needs calibration. The ambient light would be useful to locally control display backlight/brightness. I'm thinking to solder directly to the module's I2C terminals (parallel with the touch controller), as the sensor uses a different I2C address, it's perfectly feasible.
Or, at least, have a 1-wire DS18B20 instead of the analog one, soldered in place.

**Comment by fvanroie** [comment]
Hi, thanks for the interest in hasp-lvgl. In hasp-lvgl _-at the moment-_ custom sensors will require you to build a custom firmware. Adding the needed libraries and wrapper code should be straight-forward.

There are also a lot of projects that handle custom temperature sensors already, like tasmota, espHome, etc... Our focus is currently on the hardware in the supported devices and built-in sensors for the pre-built firmware files.

The potential number of sensors is huge, so I'd like to find out which particular ones would warrant built-in support. When a sensor is widely used or present in multiple touchscreen devices, we can reconsider.

I do like the idea to have support for common sensors, but for the time being it is on the back-burner.

**Reply by nagyrobi** [reply]
Other projects are out of scope, as you can't run for example hasp-lvgl and tasmota on the same device at the same time.

Look on the practical side: you'd put a Lanbon L8 with hasp-lvgl in each room to control things, but you also need reliable temperature and humidity to make a thermostat. Currently for this you need a second device, placed somewhere, which needs power, place, and a second network connection etc. just to get the temp/hum.

Can you please point me in the right direction to build my own custom version with sensors?

**Reply by fvanroie** [reply]
hasp-lvgl can be compiled with Visual Studio Code and Platformio. A small guide can be found in de [hasp-docs](https://fvanroie.github.io/hasp-docs/#compiling/) and the example for Lanbon L8 on [blakadder.com](https://blakadder.com/lanbon-L8-custom-firmware/#installing). Add the library for your sensor and try to print out the values in the main loop(). Once that is achieved, sending the values over mqtt is quite easy.

> you can't run for example hasp-lvgl and tasmota on the same device at the same time

We have tested hasp-lvgl as a Tasmota submodule. They *can* run on the same device, but that is more difficult and involves even more work. Adding a second device or compiling from hasp-lvgl source are the easiest solutions.

**Reply by nagyrobi** [reply]
Compiling from source is what I'm looking for.

**Reply by fvanroie** [reply]
Check out https://fvanroie.github.io/hasp-docs/#compiling/

**Reply by netmikey** [reply]
> We have tested hasp-lvgl as a Tasmota submodule

Interesting, reading through this topic I kept thinking that it would be awesome to have openHASP as ESPHome submodule, cause I just ordered an ESP32 TouchDown and first extensions that come to mind are temperature, humidity and ambient light sensors.

Have you considered ESPHome? I know it plays in the same ballpark as Tasmota, but it is what I know for now... I like the idea quite a bit: you would avoid having to add explicit support for any hardware extension as that would be left to ESPHome, that's its main purpose after all. Also, you could just use stuff built into ESPHome like Improv for Network credential provisioning. It just feels like you could avoid a lot of work down the road having to explicitly support all of what's already available in ESPHome.

I understand your point that it's not too complicated to add support for any sensor in VSCode/PlatformIO, but if I do that, it means I'll have to re-integrate and re-build my custom binary with each and every openHASP release, which is a bit of a bummer... Leveraging ESPHome to build my image using my config file at the press of a button would be much more convenient :-)

**Reply by fvanroie** [reply]
I have experimented with ESPHome and lvgl can run as a submodule just fine.
The problem is that openHASP relies heavily on ArduinoJSON 6 and ESPHome ships with an old *custom* ArduinoJSON 5 library... So, being dependent on another project also has some drawbacks.

For sensors, I think a custom code folder might be possible. There you can put your own functions which hook into openHASP. The folder would be kept out of version control, so you can update the project files and it will recompile with the customizations...

**Reply by netmikey** [reply]
> being dependent on another project also has some drawbacks

Yes indeed. In the end, it's a cost/benefit question.

> The problem is that openHASP relies heavily on ArduinoJSON 6 and ESPHome ships with an old custom ArduinoJSON 5 library

Oh I see. Maybe we could motivate the ESPHome project to move to ArduinoJSON 6 if that would pave the way for openHASP on ESPHome?

**Reply by fvanroie** [reply]
I have only looked into this briefly and didn't dig much deeper into it. From what I understand on the ESPHome Discord channel, that library was modified a long time ago by the original creator and now no-one really knows what was changed...

I'd be happy to revisit this when the ESPHome ArduinoJson version has been updated. I do have hopes that Nabu Casa will eventually come around to updating the library. Or if some people would take this up as a sprint. Maybe other ESPHome projects would also profit from such an update.

Maybe it's even possible to run two distinct versions side-by-side using namespaces.... but I'm not experienced in that.

**Reply by nagyrobi** [reply]
> For sensors, I think a custom code folder might be possible. There you can put your own functions which hook into openHASP. The folder would be kept out of version control, so you can update the project files and it will recompile with the customizations...

That sounds awesome! How about re-using Tasmota's xsns_*.ino sensor drivers? Each sensor type is in a different file, gets loaded from user_customize. It's already preparing the MQTT payload and the display of the value on the web interface straight from the driver.

We need it to appear in the GPIO dropdowns, so it can be assigned and bang...

**Reply by fvanroie** [reply]
There are way too many dependencies on headers and other files/libraries to have a drop-in .ino automagically work...
You could add a wrapper yourself if needed... but for now I was just thinking of proper Arduino libraries.

**Reply by fvanroie** [reply]
A proof-of-concept LVGL component for ESPhome can be found here: https://github.com/fvanroie/esphome-lvgl.

**Comment by fvanroie** [comment]
You can add `#define HASP_USE_CUSTOM 1` to your `include\user_config_override.h`.
Copy the default template files as `my_custom.h` and `my_custom.cpp`

This will add the `custom\my_custom.h` header to openHASP compilation and call the defined functions from the main loop...

**Comment by kydang789** [comment]
> Would it be possible to add software support for example SHTC3 (high precision temperature and humidity sensor) or MAX44009 (ambient light sensor) both on I2C bus. All it needs to do is publish to an MQTT topic the value of the sensors at regular time interval, say, at every 30 seconds.
> 
> I'm specifically looking at Lanbon L8. It has an analog temp sensor by factory, but that kind is quite unreliable and needs calibration. The ambient light would be useful to locally control display backlight/brightness. I'm thinking to solder directly to the module's I2C terminals (parallel with the touch controller), as the sensor uses a different I2C address, it's perfect
[EM20918_code.zip](https://github.com/HASwitchPlate/openHASP/files/7788101/EM20918_code.zip)
ly feasible. Or, at least, have a 1-wire DS18B20 instead of the analog one, soldered in place.

WT32-SC01 has this sensor. Also code for this, did you try?
[EM20918 Datasheet 2019 Oct.pdf](https://github.com/HASwitchPlate/openHASP/files/7788102/EM20918.Datasheet.2019.Oct.pdf)
[EM20918_code.zip](https://github.com/HASwitchPlate/openHASP/files/7788105/EM20918_code.zip)

**Reply by fvanroie** [reply]
> WT32-SC01 has this sensor. Also code for this, did you try?

My WT32-SC01 do not include this sensor. Which version to you have?
Can you let me know the I2C address of the sensor?

If your WT32-SC01 has this sensor included, you can add the driver code to openHASP using `my_custom.h` and `my_custom.cpp`

**Reply by kydang789** [reply]
> > WT32-SC01 has this sensor. Also code for this, did you try?
> 
> My WT32-SC01 do not include this sensor. Which version to you have? Can you let me know the I2C address of the sensor?
> 
> If your WT32-SC01 has this sensor included, you can add the driver code to openHASP using `my_custom.h` and `my_custom.cpp`

This here

https://github.com/HASwitchPlate/openHASP/files/7788102/EM20918.Datasheet.2019.Oct.pdf
https://github.com/HASwitchPlate/openHASP/files/7788105/EM20918_code.zip

![image](https://user-images.githubusercontent.com/68098745/147721428-fcb7ba69-f571-4a17-b07f-3fb2a00dce0a.png)

**Reply by fvanroie** [reply]
That sensor is _not_ populated on the PCBs of my WT32-SC01 displays...

**Reply by kquinsland** [reply]
> That sensor is _not_ populated on the PCBs of my WT32-SC01 displays...

Likewise, but I was considering populating it as being able to adjust display status / brightness is rather important for one of my deployment scenarios. I do not want 100% backlight  in a dim/dark room as that would be _super_ distracting... but when somebody approaches the display, the backlight should increase to some appropriate level.  I was also toying with the concept of a 'screensaver' that would display date/time and maybe some status info but that would be replaced with a different screen when somebody was actively infront of the display. Display the current playing song in large type until somebody walks up.. and then switch to a new screen where the track info is much smaller because the screen now also has some playback controls / volume adjust ... etc. 

Unfortunately the EM20918 does not include a raw lux level readout so I was planning on using a LDR attached an an analogue pin _or_ another sensor like`VEML7700` to do the ambient light readings.

**Reply by magicjohansson** [reply]
I have been using OPENHasp as user interfaces for my Home Assistant, running on Lanbon L8. What a great software, I really appreciate the work put into this. 
I have also been running a few ESP32-boards as interfaces for some sensors, also works great. 

The next step would be to replace all room-controllers and floor heating thermostats with L8's running OPENhasp. My old heater thermostats has 2 sensors, one for floor temp and another for room temperature. I have been trying to make custom softare for OPENhasp to read the temperatures off 2 DS18B20 sensors but with no success. I disconnected the pins for relay K3 and conntected the OneWire bus to that pin. I can control the pin ON and OFF if I set it to OUTPUT for test but when I try to use the OneWire and dallastemperature libraries it does not detect any sensors.

I know the question is quite broad but I still hope someone has some input on this. Have You for example had success running OneWire or other libraries on the L8 ?

My custom code is here below

```
/* MIT License - Copyright (c) 2019-2022 Francis Van Roie
   For full license information read the LICENSE file in the project folder */

// USAGE: - Copy this file and rename it to my_custom.cpp
//        - Change false to true on line 9

#include "hasplib.h"

#if defined(HASP_USE_CUSTOM) && true // <-- set this to true in your code

#include "hasp_debug.h"
#include "custom/my_custom.h"
#include "OneWire.h" 
#include "DallasTemperature.h"

// Data wire is plugged into pin 12 on the Lanbon L8 
#define ONE_WIRE_BUS 12 
#define MoodGreen 32

// Setup a oneWire instance to communicate with any OneWire devices  
OneWire oneWire(ONE_WIRE_BUS); 
/********************************************************************/
// Pass our oneWire reference to Dallas Temperature. 
DallasTemperature sensors(&oneWire);
int NumberOfSensors;
float temperatureC[10];

void custom_setup()
{
  // Initialize onewire sensors
  sensors.begin();
  // Set pin for grren moodlight as output  
  pinMode(MoodGreen, OUTPUT);
}

void custom_loop()
{
    // Non-blocking code here, this should execute very fast!
}

void custom_every_second()
{
}

void custom_every_5seconds()
{
  int i;

//  LOG_VERBOSE(TAG_CUSTOM, "5 seconds have passsed...");
//  dispatch_state_subtopic("my_sensor","{\"test\":123}");

  digitalWrite(MoodGreen, !digitalRead(MoodGreen));

  sensors.begin();
  NumberOfSensors = sensors.getDeviceCount();
  Serial.print("Found ");
  Serial.print(NumberOfSensors, DEC);
  Serial.println(" sensors.");

  sensors.requestTemperatures(); // Send the command to get temperature readings 
  if (NumberOfSensors > 0) {
    for (i = 0; i<NumberOfSensors;i++) {
      temperatureC[i] = sensors.getTempCByIndex(i);
      Serial.print(temperatureC[i]);
      Serial.print("  ");
    } 
    Serial.println("ºC");
  }
}

bool custom_pin_in_use(uint8_t pin)
{
  if(pin == ONE_WIRE_BUS) return true; 
  if(pin == MoodGreen) return true; 
  // otherwise the pin is not used
  return false;
}

void custom_get_sensors(JsonDocument& doc)
{
    /* Sensor Name */
    JsonObject sensor = doc.createNestedObject(F("Custom"));

    /* Key-Value pair of the sensor value */
    sensor[F("Random")] = random(256);
}

void custom_topic_payload(const char* topic, const char* payload, uint8_t source){
    // Not used
}

#endif // HASP_USE_CUSTOM

```

**Reply by magicjohansson** [reply]
Good morning

after a good nights sleep and more investigation I can answer this myself. 
The GPIO's for K1, K2 and K3 are not directly connected to the pins on the connector to the relay board but there are driver transistors in between. hence I can control the putput on/off and see it on the pin but of course the 2-direction communication is not working.

I will continue investigating how big changes are needed to get this to work.

To be continued . . . . .

**Reply by nagyrobi** [reply]
Let us know as I'm very interested in your findings.

**Reply by nagyrobi** [reply]
You could re-use the RX and TX pins, but that also requires disabling the serial console (which is pretty useless anyway when mounted in the wall). @fvanroie is it enough to set `#define HASP_START_CONSOLE 0` to free up these pins?

**Reply by nagyrobi** [reply]
I just finished doing calibration measurements on a batch of DS18B20 and SHTC3 and must say I'm impressed of the second:
https://a.aliexpress.com/_mrhJsHO

How difficult would it be to remap RX and TX (in Lanbon) as a secondary I2C bus?

**Reply by fvanroie** [reply]
`HASP_START_CONSOLE ` only prevents reading from `RX`. `TX` will still output the logs.
`HASP_USE_SERIAL` would be needed to turn off serial communications completely...

**Reply by nagyrobi** [reply]
That will free up the RX/TX pins?

**Reply by fvanroie** [reply]
possibly... `HASP_USE_SERIAL` is just a suggestion, it's not available at this point.
Serial is always active atm, whether you need it or not.

**Comment by samturner3** [comment]
@nagyrobi how did you go with this? I am in same boat, I want to add I2C devices to Lanbon L8.

Also coding in `my_custom.h` and `my_custom.cpp` solder directly to the module's I2C terminals (parallel with the touch controller).
Only thing I can not think of is how to hook into existing I2C bus used for touch controller, rather than initiate a new from a lib we include ourselves, as obviously it is already running from stock firmware for the screen.

Found these functions in file `include/lv_drv_conf.h` from stock firmware.
![Screen Shot 2022-07-05 at 4 14 59 pm](https://user-images.githubusercontent.com/14887388/177261862-fbc17d84-1e47-452e-a2fb-9ae8d5753aed.png)

I'm guessing we somehow need to access these from our custom code?

(I have not tried anything yet)
I can't see how to access that instance from just the `custom_*` functions.

If you did do it, do you have code example you could share?
Or @fvanroie have any suggestions?

If not doable via custom code, maybe a modification to the firmware (slightly out of my depth); or ideally exposing these so it can be done via custom code?

**Reply by nagyrobi** [reply]
I gave up on this idea, went with Xiaomi LYWSD03MMC running pvvx firmware, much better approach system-wise, as you can place the sensors anywhere - in optimal measurement locations - which usually can't be done with a touch screen.

Pricewise they are attractive and I can receive the signal of 10 such sensors from all around the house via a single, Ethernet-based ESPHome node [reliably](https://github.com/nagyrobi/home-assistant-configuration-examples/blob/main/esphome/ble-sensors-ethernet.yaml).
Battery seems to last long indeed, I run them for 5 months now, and currently at 75-80%.

**Reply by fvanroie** [reply]
I think you can just use `#include <Wire.h>` and then use the global `Wire` object with the appropriate address:
```cpp
    do {
        Wire.beginTransmission(I2C_ADDR); 
        Wire.write(addr); 
        Wire.endTransmission(false); // Restart
        delay(10); 
        rdDataCount = Wire.requestFrom(I2C_ADDR, 1); 
    } while(rdDataCount == 0); 
    while(Wire.available()) {
        rdData = Wire.read(); 
    }
```
The touch driver initializes the bus in `src\drv\touch\touch_driver_ft6336u.h`. You should be able to use it in the custom code too.

If it doesn't work, you could use `Wire1` on separate pins to initialize the secondary I²C bus with your own config.

**Reply by flexiti** [reply]
I am using ESP32-4848S040  (Does anyone have schematic ?)
I think the advice  @fvanroie  is ok, in this device you need to use the pin SDA =19 and SCL = 45
The transmission speed is set to 400000
I need to check whether this I2C bus works at 3V or 5V logic
Has anyone managed to connect something in parallel with G911 touch sensor??
I will try to connect BMP280 or BME280
```

-D TOUCH_DRIVER=0x911
    -D TOUCH_WIDTH=480
    -D TOUCH_HEIGHT=480
    -D TOUCH_SDA=19
    -D TOUCH_SCL=45
    -D TOUCH_RST=-1
    -D TOUCH_IRQ=-1
    -D I2C_TOUCH_FREQUENCY=400000
    -D I2C_TOUCH_ADDRESS=0x5D ; or 0x14
    -D I2C_TOUCH_PORT=1
```

**Comment by flexiti** [comment]
https://github.com/HASwitchPlate/openHASP/discussions/666
